libr/core/windows_heap.c

```diff
@@ -0,0 +1,822 @@
+*		1) Calling InitHeapInfo with both PDI_HEAPS and PDI_HEAP_BLOCKS.
+*			This will fill a buffer with HeapBlockBasicInfo like structures which
+*			is then walked through by calling GetFirstHeapBlock and subsequently GetNextHeapBlock
+*			(see 1st link). This approach is the more generic one as it uses Windows functions.
+*			Unfortunately it fails to offer more detailed information about each block (although it is possible to get this info later) and
+*			also fails misteriously once the count of allocated blocks reach a certain threshold (1mil or so) or if segment heap is active for the
+*			program (in this case everything locks in the next call for the function)
+*		2) In case 1 fails, Calling GetHeapBlocks, which will manually read and parse (poorly :[ ) each block.
+*			First it calls InitHeapInfo	with only the PDI_HEAPS flag, with the only objective of getting a list of heap header addresses. It will then
+*			do the job that InitHeapInfo would do if it was called with PDI_HEAP_BLOCKS as well, filling a buffer with HeapBlockBasicInfo structures that
+*			can also be walked with GetFirstHeapBlock and GetNextHeapBlock (and HeapBlockExtraInfo when needed).
+*
+*	TODO:
+*		Var to select algorithm?
+*		x86 vs x64 vs WOW64
+*		Graphs
+*		Use the flags field for extra info
+*		Print structures
+*		Make sure GetHeapBlocks actually works
+*		Maybe instead of using hardcoded structs we can get the offsets from ntdll.pdb
+*/
+
+#define PDI_MODULES				0x01
+#define PDI_HEAPS				0x04
+#define PDI_HEAP_TAGS			0x08
+#define PDI_HEAP_BLOCKS			0x10
+#define PDI_HEAP_ENTRIES_EX		0x200
+
+#pragma comment (lib, "ntdll.lib")
+
+#define CHECK_INFO(heapInfo)											\
+	if (!heapInfo) {													\
+		eprintf ("It wasn't possible to get the heap information\n");	\
+		return;															\
+	}																	\
+	if (!heapInfo->count) {												\
+		r_cons_print ("No heaps for this process\n");					\
+		return;															\
+	}																	\
+
+#define UPDATE_FLAGS(hb, flags)					\
+	if ((flags & 0xf1) || (flags & 0x0200)) {	\
+		hb->dwFlags = LF32_FIXED;				\
+	} else if ((flags & 0x20)) {				\
+		hb->dwFlags = LF32_MOVEABLE;			\
+	} else if ((flags & 0x0100)) {				\
+		hb->dwFlags = LF32_FREE;				\
+	}											\
+
+static bool is_segment_heap(HANDLE h_proc, PVOID heapBase) {
+	HEAP heap;
+	if (ReadProcessMemory (h_proc, heapBase, &heap, sizeof (HEAP), NULL)) {
+		if (heap.SegmentSignature == 0xddeeddee) {
+			return true;
+		}
+	}
+	return false;
+}
+
+// These functions are basically Heap32First and Heap32Next but faster
+static bool GetFirstHeapBlock(PDEBUG_HEAP_INFORMATION heapInfo, PHeapBlock hb) {
+	r_return_val_if_fail (heapInfo && hb, false);
+	PHeapBlockBasicInfo block;
+
+	hb->index = 0;
+	hb->dwAddress = 0;
+	hb->dwFlags = 0;
+	hb->extraInfo = NULL;
+
+	block = (PHeapBlockBasicInfo)heapInfo->Blocks;
+	SIZE_T index = hb->index;
+	do {
+		hb->dwAddress = (void *)block[index].address;
```

Comment:
unchecked array access?

Reply:
here and other places

Reply:
I added some checks and I'm always calling `GROW_BLOCKS()` before writing to the array, so it should be ok now.